pkg/virt-controller/watch/drain/evacuation/evacuation.go

```diff
@@ -0,0 +1,442 @@
+
+	if diff == 0 {
+		// nothing to do
+		return nil
+	}
+
+	// TODO: should the order be randomized?
+	selectedCandidates := migrationCandidates[0:diff]
+	log.DefaultLogger().Infof("node: %v, migrations: %v, candidates: %v, selected: %v", node.Name, len(activeMigrations), len(migrationCandidates), len(selectedCandidates))
+
+	wg := &sync.WaitGroup{}
+	wg.Add(diff)
+
+	errChan := make(chan error, diff)
+
+	c.migrationExpectations.ExpectCreations(node.Name, diff)
+	for _, vmi := range selectedCandidates {
+		go func(vmi *virtv1.VirtualMachineInstance) {
+			defer wg.Done()
+			_, err := c.clientset.VirtualMachineInstanceMigration(vmi.Namespace).Create(&virtv1.VirtualMachineInstanceMigration{
+				ObjectMeta: v1.ObjectMeta{
+					GenerateName: "kubevirt-evacuation-",
+				},
+				Spec: virtv1.VirtualMachineInstanceMigrationSpec{
+					VMIName: vmi.Name,
+				},
+			})
+			if err != nil {
+				c.migrationExpectations.CreationObserved(node.Name)
+				errChan <- err
+			}
+		}(vmi)
+	}
+
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+	}
+	return nil
+}
+
+func (c *EvacuationController) migrationForVMI(namespace, name string) (*virtv1.VirtualMachineInstanceMigration, error) {
+	migrations, err := c.migrationInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, migration := range migrations {
+		p := v1.GetControllerOf(migration.(*virtv1.VirtualMachineInstanceMigration))
+		if p != nil && p.Kind == virtv1.VirtualMachineInstanceGroupVersionKind.Kind &&
+			p.Name == name {
+			return migration.(*virtv1.VirtualMachineInstanceMigration), nil
+		}
+	}
+	return nil, nil
+}
+
+func (c *EvacuationController) listVMIsOnNode(nodeName string) ([]*virtv1.VirtualMachineInstance, error) {
+	objs, err := c.vmiInformer.GetIndexer().ByIndex("node", nodeName)
+	if err != nil {
+		return nil, err
+	}
+	vmis := []*virtv1.VirtualMachineInstance{}
+	for _, obj := range objs {
+		vmis = append(vmis, obj.(*virtv1.VirtualMachineInstance))
+	}
+	return vmis, nil
+}
+
+func (c *EvacuationController) listRunningMigrations() ([]*virtv1.VirtualMachineInstanceMigration, error) {
+	objs := c.migrationInformer.GetStore().List()
+	migrations := []*virtv1.VirtualMachineInstanceMigration{}
+	for _, obj := range objs {
+		migration := obj.(*virtv1.VirtualMachineInstanceMigration)
+		if migration.Status.Phase != virtv1.MigrationFailed && migration.Status.Phase != virtv1.MigrationSucceeded {
+			migrations = append(migrations, migration)
+		}
+	}
+	return migrations, nil
+}
+
+func (c *EvacuationController) filterNonMigratingVMIs(vmis []*virtv1.VirtualMachineInstance, migrations []*virtv1.VirtualMachineInstanceMigration) []*virtv1.VirtualMachineInstance {
+	lookup := map[string]bool{}
+	for _, migration := range migrations {
+		lookup[migration.Namespace+"/"+migration.Spec.VMIName] = true
+	}
+
+	migrationCandidates := []*virtv1.VirtualMachineInstance{}
+
+	for _, vmi := range vmis {
```

Comment:
I guess that will change with #2059? Would make sense to see based on checking the existing migration objects how many migrations are active.

Reply:
I don't know if that would change. A canceled migration will be completed and have EndTimestamp. Until then it is still active, even if Aborted.

Reply:
There will be no conditions or so which will indicate that a VMI was not just aborted but also stopped? It might be great if higher-level controllers can do decisions based on migration objects and while looking at parts of the cluster (e.g. only vmis on some nodes) and do decisions, instead of re-interpreting this pretty low-level status and having to look it up on every VMI in the cluster.

Reply:
migrations have now a finalizer with #2059. I would prefer to stay on migration objects.